salsa-2022-tests/tests/builder_api.rs

```diff
@@ -0,0 +1,65 @@
+#![allow(warnings)]
+
+use expect_test::expect;
+use salsa::Durability;
+#[salsa::jar(db = Db)]
+struct Jar(MyInput, MySingletonInput);
+
+trait Db: salsa::DbWithJar<Jar> {}
+
+#[salsa::input(jar = Jar)]
+struct MyInput {
+    field1: u32,
+    field2: u32,
+}
+
+#[salsa::input(singleton)]
+struct MySingletonInput {
+    field1: u32,
+}
+
+#[salsa::db(Jar)]
+#[derive(Default)]
+struct Database {
+    storage: salsa::Storage<Self>,
+}
+
+impl salsa::Database for Database {}
+
+impl Db for Database {}
+
+#[test]
+fn test_builder() {
+    let mut db = Database::default();
+    // durability set with new is always low
+    let mut input = MyInput::new(&db, 12, 13);
+    input
+        .set_field1(&mut db)
+        .with_durability(Durability::HIGH)
+        .to(20);
+    input
+        .set_field2(&mut db)
+        .with_durability(Durability::HIGH)
+        .to(40);
+    let input_from_builder = MyInput::new_builder()
+        .with_durability(Durability::HIGH)
+        .with_fields(20, 40)
+        .build(&db)
+        .unwrap();
```

Comment:
Returning a `Result` from `build` isn't useless, but I also feel like almost always having to `unwrap` it will be something very repetitive and rarely useful. If you agree with my suggestion that the durability should have a default I have a different idea, which I think will make the API better, instead of having the `build` method be the terminal one, we could instead move that over to `with_fields` and we'd no longer need to return a `Result`, because we know that we have the values and the other thing we need is the db, but there a panic would be fine.
```rust    
let input_from_builder = MyInput::new_builder()
        .with_db(&db)
        .with_durability(Durability::HIGH)
        .with_fields(20, 40);
```
I think this makes the API calls shorter, with no downsides and it would be closer to the `salsa::Setter` API. The only thing that I kind of don't like is how to fields method is named similarly to the other ones, which doesn't clearly show that it's the terminal method that actually builds the type.

Reply:
> The only thing that I kind of don't like is how to fields method is named similarly to the other ones, which doesn't clearly show that it's the terminal method that actually builds the type.

That's a real concern. I actually think builders should have `build`, and should naturally return `Result`s except when the value we're building can comfortably have default values all through. In this case, `field`s cannot have default values. And, panicking inside `with_fields`? hmmn maybe people don't expect that. But I wouldn't know. I don't have much experience here. If you're so sure...

Maybe others have something to say?
@XFFXFF @nikomatsakis @crlf0710

Reply:
> In this case, fields cannot have default values.

That and the database are the two things that we need to set for the struct to be complete. I think either one of those could be used as the `build` method and panic only if the other is missing, but it's a tradeoff where we make the API calls slightly shorter, but lose a bit of clarity of which method is the one that `build`s the type.
But panicking instead of returning `Result` and making either `with_db` or `with_fields` the `build` method are two separate things. If you're sure that the `build` should return `Result` you can still consider putting that logic in `with_db` or `with_fields`.

> But I wouldn't know. I don't have much experience here. If you're so sure...

I don't either, so take what I say with a grain of salt. I just wanted to bring it up as a talking point, so we could discuss it with the others.

Reply:
Coming back to this with fresh eyes, a few thoughts--

1. I think that we should strive to avoid `Result` that are statically known to not apply. 
2. I don't love having `with_values` take a tuple, I think I'd rather have a method per field.
3. This seems to imply we may need to do "type tricks" to make a builder that is statically known to have a value for all fields etc.

It seems like we should back up from the PR and first we be more specific about the API that we want for an example.